src/R8R.sol

```diff
@@ -0,0 +1,170 @@
+
+    // ==============
+    // === ERRORS ===
+    // ==============
+    error Error__TokenTransferFailed();
+
+    // ===================
+    // === CONSTRUCTOR ===
+    // ===================
+
+    constructor(
+        address _robot,
+        uint256 _gameEntryPriceInEth,
+        IERC20[] memory _gameTokensAllowedList,
+        uint256[] memory _gameTokensEntryPriceList
+    ) Ownable(msg.sender) {
+        robot = _robot;
+        gameEntryPriceInEth = _gameEntryPriceInEth;
+        gameTokensAllowedList = _gameTokensAllowedList;
+        gameTokensEntryPriceList = _gameTokensEntryPriceList;
+    }
+
+    // =========================
+    // === PUBLIC FUNCTIONS ====
+    // =========================
+
+    // @notice called by AI wallet to initialise a new game
+    // @notice creates a new Game struct that players + ratings can then be added to when they join
+    function createGame() public onlyRobot {
+        // increment gameId
+        gameId = gameId + 1;
+
+        // empty placeholder arrays for players + their ratings (until someone joins the game)
+        address[] memory playersArray;
+        uint256[] memory playersRatingsArray;
+
+        // create game
+        Game memory game = Game({
+            gameId: gameId,
+            players: playersArray,
+            playerRatings: playersRatingsArray,
+            robotRating: 0,
+            winner: address(0), // set this to our treasury address upon initial game creation??
+            winnerPayout: 0
+        });
+
+        // add new game to game mappings indexed by gameId
+        games[gameId] = game;
+
+        // emit event
+        emit GameCreated(gameId);
+    }
+
+    // @notice allows new players to join a selected game
+    // @param user selects a _gameId to play & submits a rating between 1 - 100 for the image
+    function joinGameWithEth(uint256 _gameId, uint256 _playerRating) public payable nonReentrant {
+        // Checks
+        require(msg.value > 0, "Please send Eth to enter");
+        require(msg.value == gameEntryPriceInEth, "Please send correct amount of Eth to enter");
+        require(_gameId > 0 && _gameId <= gameId, "Please select a game to enter");
+        require(_playerRating > 0 && _playerRating < 101, "Please provide a rating of between 1 - 100");
+
+        // Effects
+        // Add msg.sender to players array & _playerRating to playerRatings array
+        games[_gameId].players.push(msg.sender);
+        games[_gameId].playerRatings.push(_playerRating);
+
+        // Interactions - none, Eth accepted into contract
+
+        // emit event
+        emit PlayerJoinedGameWithEth(msg.sender, _gameId);
+    }
+
+    // @notice allows new players to join a selected game using ERC20 tokens
+    // @param players must pass in an amount of an allow-listed token along with their selected _gameId & image rating
+    function joinGameWithTokens(IERC20 _token, uint256 _amountOfToken, uint256 _gameId, uint256 _playerRating)
```

Comment:
We can likely combine this function with the above function and just implement some conditional logic to have the function understand which is trying to be used.